cache/lru_secondary_cache.cc

```diff
@@ -0,0 +1,171 @@
+    const Slice& key, const Cache::CreateCallback& create_cb, bool /*wait*/) {
+  std::unique_ptr<SecondaryCacheResultHandle> handle;
+  Cache::Handle* lru_handle = cache_->Lookup(key);
+  if (lru_handle == nullptr) {
+    return handle;
+  }
+
+  void* value = nullptr;
+  size_t charge = 0;
+  Status s;
+
+  if (cache_options_.compression_type == kNoCompression) {
+    s = create_cb(cache_->Value(lru_handle), cache_->GetCharge(lru_handle),
+                  &value, &charge);
+  } else {
+    UncompressionContext uncompression_context(cache_options_.compression_type);
+    UncompressionInfo uncompression_info(uncompression_context,
+                                         UncompressionDict::GetEmptyDict(),
+                                         cache_options_.compression_type);
+
+    size_t uncompressed_size = 0;
+    CacheAllocationPtr uncompressed;
+    uncompressed =
+        UncompressData(uncompression_info, (char*)cache_->Value(lru_handle),
+                       cache_->GetCharge(lru_handle), &uncompressed_size,
+                       cache_options_.compress_format_version);
+
+    if (!uncompressed) {
+      return handle;
+    }
+    s = create_cb(uncompressed.get(), uncompressed_size, &value, &charge);
+  }
+
+  if (!s.ok()) {
+    return handle;
+  }
+
+  handle.reset(new LRUSecondaryCacheResultHandle(cache_.get(), lru_handle,
+                                                 value, charge));
+
+  return handle;
+}
+
+Status LRUSecondaryCache::Insert(const Slice& key, void* value,
+                                 const Cache::CacheItemHelper* helper) {
+  size_t size = (*helper->size_cb)(value);
+  // TODO optimization.
+  std::unique_ptr<char[]> buf{new char[size]};
+  if (cache_options_.memory_allocator) {
+    CacheAllocationPtr ptr =
+        AllocateBlock(size, cache_options_.memory_allocator.get());
+    buf.reset(ptr.get());
+  }
+
+  Status s = (*helper->saveto_cb)(value, 0, size, buf.get());
+  if (!s.ok()) {
+    // delete[] buf;
+    return s;
+  }
+  Slice val(buf.get(), size);
+
+  std::string compressed_val;
+  if (cache_options_.compression_type != kNoCompression) {
+    CompressionOptions compression_opts;
+    CompressionContext compression_context(cache_options_.compression_type);
+    uint64_t sample_for_compression = 0;
+    CompressionInfo compression_info(
+        compression_opts, compression_context, CompressionDict::GetEmptyDict(),
+        cache_options_.compression_type, sample_for_compression);
+
+    bool success =
+        CompressData(val, compression_info,
+                     cache_options_.compress_format_version, &compressed_val);
+
+    if (!success) {
+      // delete[] buf;
+      return Status::Corruption("Error compressing value.");
+    }
+
+    val = Slice(compressed_val);
+  }
+
+  std::unique_ptr<char[]> copy{new char[val.size()]};
+  memcpy(copy.get(), val.data(), val.size());
+
+  s = cache_->Insert(key, copy.get(), val.size(), helper->del_cb);
```

Comment:
We can't pass helper->del_cb here, as that's for a Block object. I believe what we need to do here is ```new CacheAllocationPtr```, and ```std::move``` either ```copy``` or ```buf``` to it (depending on compressed or uncompressed). Then pass a pointer to the allocated CacheAllocatorPtr as value, and a deleter function that casts the input argument to a CacheAllocatorPtr* and destructs it.